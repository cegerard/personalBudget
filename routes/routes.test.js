const slugify = require('slugify');
const request = require('supertest');

const ExpenseModelStub = require('../test/stubs/ExpenseModelStub');
const BudgetModelStub = require('../test/stubs/BudgetModelStub');
const app = require('../app').app;
const ExpenseRepository = require('../data').ExpenseRepository;
const BudgetRepository = require('../data').BudgetRepository;

describe('Route', () => {
  let budgetRepository;
  let expenseRepository;
  beforeAll(() => {
    BudgetModelStub.resetStore();
    ExpenseModelStub.resetStore();
    budgetRepository = new BudgetRepository(BudgetModelStub);
    expenseRepository = new ExpenseRepository(ExpenseModelStub);
  });

  describe('GET /notFound', () => {
    it('should response with 404', async () => {
      await request(app).get('/notFound').expect(404);
    });
  });

  describe('GET /', () => {
    it('should response with 302', async () => {
      await request(app).get('/').expect(302);
    });

    it('should render home page', async () => {
      await request(app)
        .get('/')
        .then((res) => {
          expect(res.text).toMatchSnapshot();
        });
    });
  });

  describe('GET /budgets', () => {
    it('should response with 200', async () => {
      await request(app).get('/budgets').expect(200);
    });

    it('should render budgets page', async () => {
      await request(app)
        .get('/budgets')
        .then((res) => {
          expect(res.text).toMatchSnapshot();
        });
    });
  });

  describe('GET /expenses', () => {
    it('should response with 200', async () => {
      await request(app).get('/expenses').expect(200);
    });

    it('should render expenses page', async () => {
      await request(app)
        .get('/expenses')
        .then((res) => {
          expect(res.text).toMatchSnapshot();
        });
    });
  });

  describe('GET /expenses/filter', () => {
    it('should response with 200', async () => {
      await request(app).get('/expenses/filter?budgetName=Essence').expect(200);
    });

    it('should render expenses filtered page', async () => {
      await request(app)
        .get('/expenses/filter?budgetName=Essence')
        .then((res) => {
          expect(res.text).toMatchSnapshot();
        });
    });
  });

  describe('POST /budgets', () => {
    const newBudget = {
      name: 'budget Name',
      amount: 42,
      description: 'little description',
    };

    it('should create a new budget line', async () => {
      await request(app)
        .post('/budgets')
        .set('Content-Type', 'application/json')
        .send(newBudget)
        .expect(200)
        .then(async () => {
          const budgets = await budgetRepository.find();
          const budgetsFound = budgets.filter((budget) => {
            return budget.name === newBudget.name;
          });
          expect(budgetsFound.length).toEqual(1);
          expect(budgetsFound[0]).toMatchObject({
            _id: expect.any(String),
            name: newBudget.name,
            slug: slugify(newBudget.name),
            amount: newBudget.amount,
            description: newBudget.description,
            available: newBudget.amount,
            expenses: [],
          });
        });
    });
  });

  describe('POST /expenses', () => {
    const newExpense = {
      name: 'test expense',
      amount: 12,
      date: '2020-09-10',
      budgetlineId: '1',
    };

    it('should create a new expense', async () => {
      await request(app)
        .post('/expenses')
        .set('Content-Type', 'application/json')
        .send(newExpense)
        .expect(200)
        .then(async () => {
          const expenses = await expenseRepository.find();
          const expensesFound = expenses.filter((expense) => {
            return expense.name === newExpense.name;
          });
          expect(expensesFound.length).toEqual(1);
          expect(expensesFound[0]).toMatchObject({
            _id: expect.any(String),
            name: newExpense.name,
            amount: newExpense.amount,
            date: newExpense.date,
            budgetLine: {
              _id: newExpense.budgetlineId,
              name: 'ElectricitÃ©',
            },
          });
        });
    });

    // TODO add error mangement
    // it('should return 404 if budget not found', async () => {
    //   newExpense.budgetlineId = 'notfound';
    //   await request(app)
    //     .post('/expenses')
    //     .set('Content-Type', 'application/json')
    //     .send(newExpense)
    //     .expect(404);
    // });
  });

  describe('DELETE /expenses', () => {
    it('should remove expense from expense list and the budget line', async () => {
      await request(app)
        .delete('/expenses/100')
        .expect(204)
        .then(async () => {
          const allExpenses = await expenseRepository.find();
          const expenseNotFound = allExpenses.find((expense) => {
            return expense._id === 100;
          });

          expect(expenseNotFound).toBeUndefined();

          const allBudgets = await budgetRepository.find();
          const budgetList = allBudgets.find((budget) => {
            return budget._id === '4';
          });
          const expenseNotFoundInBudgetLine = budgetList.expenses.find((expense) => {
            return expense._id === 100;
          });

          expect(expenseNotFoundInBudgetLine).toBeUndefined();
        });
    });

    it('should return a 404 error when the expense can not be deleted', async () => {
      await request(app).delete('/expenses/404').expect(404);
    });
  });

  describe('DELETE /budgets', () => {
    it('should remove budget line and all its expenses', async () => {
      await request(app)
        .delete('/budgets/4')
        .expect(204)
        .then(async () => {
          const allBudgets = await budgetRepository.find();
          const budgetNotFound = allBudgets.find((budget) => {
            return budget._id === '4';
          });

          expect(budgetNotFound).toBeUndefined();

          const allExpenses = await expenseRepository.find();
          const expensesFound = allExpenses.filter((expense) => {
            return expense.budgetLine._id === '4';
          });

          expect(expensesFound.length).toEqual(0);
        });
    });

    it('should return a 404 error when the budget can not be deleted', async () => {
      await request(app).delete('/budgets/404').expect(404);
    });
  });
});
