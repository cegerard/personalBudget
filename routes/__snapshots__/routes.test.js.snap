// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route GET / should render home page 1`] = `"Found. Redirecting to /budgets"`;

exports[`Route GET /budgets should render budgets page 1`] = `"<!DOCTYPE html><html><head><title>Personal Budget</title><link rel=\\"stylesheet\\" href=\\"/stylesheets/style.css\\"><link rel=\\"stylesheet\\" href=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.0/css/bootstrap.min.css\\" integrity=\\"sha384-9aIt2nRpC12Uk9gS9baDl411NQApFmC26EwAOH8WgZl5MYYxFfc+NcPb1dKGj7Sk\\" crossorigin=\\"anonymous\\"><link rel=\\"stylesheet\\" href=\\"https://use.fontawesome.com/releases/v5.7.0/css/all.css\\" crossorigin=\\"anonymous\\"></head><body class=\\"container\\"><nav class=\\"navbar navbar-expand-lg navbar-dark bg-dark\\"><a class=\\"navbar-brand\\" href=\\"#\\">Personal Budget</a><button class=\\"navbar-toggler\\" type=\\"button\\" data-toggle=\\"collapse\\" data-target=\\"#navbarSupportedContent\\" aria-controls=\\"navbarSupportedContent\\" aria-expanded=\\"false\\" aria-label=\\"Toggle navigation\\"><span class=\\"navbar-toggler-icon\\"></span></button><div class=\\"collapse navbar-collapse\\" id=\\"navbarSupportedContent\\"><ul class=\\"navbar-nav mr-auto\\"><li class=\\"nav-item active\\"><a class=\\"nav-link\\" href=\\"/budgets\\">Budgets</a></li><li class=\\"nav-item\\"><a class=\\"nav-link\\" href=\\"/expenses\\">Expenses</a></li></ul></div></nav><div><h1 class=\\"mb-4 mt-3\\">Budgets<button class=\\"btn btn-primary addBtn\\" type=\\"button\\" data-target=\\"#addBudgetModal\\" data-toggle=\\"modal\\"><span class=\\"fas fa-plus\\"></span></button></h1></div><div class=\\"row\\"><div class=\\"col-md-4 mb-4\\"><div class=\\"card\\"><div class=\\"card-body d-flex flex-column\\"><h5 class=\\"card-title\\">Electricité</h5><div class=\\"d-flex flex-row justify-content-between\\"><p>Amount:</p><p class=\\"align-self-end\\">58/58 €</p></div><p class=\\"card-text\\">Dépenses liées au contrat d'électricité de la maison</p><a class=\\"btn btn-primary align-self-end\\" href=\\"/budgets/1\\">Details</a></div></div></div><div class=\\"col-md-4 mb-4\\"><div class=\\"card\\"><div class=\\"card-body d-flex flex-column\\"><h5 class=\\"card-title\\">Crédit voiture</h5><div class=\\"d-flex flex-row justify-content-between\\"><p>Amount:</p><p class=\\"align-self-end\\">369.89/369.89 €</p></div><p class=\\"card-text\\">Dépenses liées au crédit des voitures</p><a class=\\"btn btn-primary align-self-end\\" href=\\"/budgets/2\\">Details</a></div></div></div><div class=\\"col-md-4 mb-4\\"><div class=\\"card\\"><div class=\\"card-body d-flex flex-column\\"><h5 class=\\"card-title\\">Crédit immobilier</h5><div class=\\"d-flex flex-row justify-content-between\\"><p>Amount:</p><p class=\\"align-self-end\\">857.12/857.12 €</p></div><p class=\\"card-text\\">Dépenses liées au crédit de la maison</p><a class=\\"btn btn-primary align-self-end\\" href=\\"/budgets/3\\">Details</a></div></div></div><div class=\\"col-md-4 mb-4\\"><div class=\\"card\\"><div class=\\"card-body d-flex flex-column\\"><h5 class=\\"card-title\\">Essence</h5><div class=\\"d-flex flex-row justify-content-between\\"><p>Amount:</p><p class=\\"align-self-end\\">79/150 €</p></div><p class=\\"card-text\\">Dépenses concernant l'essence des voitures</p><a class=\\"btn btn-primary align-self-end\\" href=\\"/budgets/4\\">Details</a></div></div></div><div class=\\"col-md-4 mb-4\\"><div class=\\"card\\"><div class=\\"card-body d-flex flex-column\\"><h5 class=\\"card-title\\">Assurance voiture</h5><div class=\\"d-flex flex-row justify-content-between\\"><p>Amount:</p><p class=\\"align-self-end\\">127/127 €</p></div><p class=\\"card-text\\">Assurance des voitures</p><a class=\\"btn btn-primary align-self-end\\" href=\\"/budgets/5\\">Details</a></div></div></div><div class=\\"col-md-4 mb-4\\"><div class=\\"card\\"><div class=\\"card-body d-flex flex-column\\"><h5 class=\\"card-title\\">Alimentation</h5><div class=\\"d-flex flex-row justify-content-between\\"><p>Amount:</p><p class=\\"align-self-end\\">500/500 €</p></div><p class=\\"card-text\\">Dépenses concernant l'alimentation (boucher, laitier, primeur et supermarché)</p><a class=\\"btn btn-primary align-self-end\\" href=\\"/budgets/6\\">Details</a></div></div></div><div class=\\"col-md-4 mb-4\\"><div class=\\"card\\"><div class=\\"card-body d-flex flex-column\\"><h5 class=\\"card-title\\">Animaux</h5><div class=\\"d-flex flex-row justify-content-between\\"><p>Amount:</p><p class=\\"align-self-end\\">120/120 €</p></div><p class=\\"card-text\\">Dépenses pour les chats</p><a class=\\"btn btn-primary align-self-end\\" href=\\"/budgets/7\\">Details</a></div></div></div><div class=\\"col-md-4 mb-4\\"><div class=\\"card\\"><div class=\\"card-body d-flex flex-column\\"><h5 class=\\"card-title\\">Enfants</h5><div class=\\"d-flex flex-row justify-content-between\\"><p>Amount:</p><p class=\\"align-self-end\\">0/0 €</p></div><p class=\\"card-text\\"></p><a class=\\"btn btn-primary align-self-end\\" href=\\"/budgets/8\\">Details</a></div></div></div><div class=\\"modal fade\\" id=\\"addBudgetModal\\"><div class=\\"modal-dialog\\"><div class=\\"modal-content\\"><div class=\\"modal-header\\"><h4 class=\\"modal-title\\">New Budget Line</h4></div><div class=\\"modal-body\\"><form method=\\"POST\\" action=\\"/budgets\\"><div class=\\"form-group\\"><label for=\\"budgetName\\">Name</label><input class=\\"form-control\\" id=\\"budgetName\\" name=\\"name\\" type=\\"text\\" aria-describedby=\\"budgetNameHelp\\" required><small class=\\"form-text text-muted\\" id=\\"budgetNameHelp\\">Budget line name must be unique</small><label for=\\"budgetAmount\\">Amount (€)</label><input class=\\"form-control\\" id=\\"budgetAmount\\" name=\\"amount\\" type=\\"number\\" step=\\"0.01\\" aria-describedby=\\"budgetAmountHelp\\" required><small class=\\"form-text text-muted\\" id=\\"budgetAmountHelp\\">Budget line amount must be gretter than 0</small><label for=\\"budgetDesc\\">Description</label><textarea class=\\"form-control\\" id=\\"budgetDesc\\" name=\\"description\\"></textarea></div><div class=\\"modal-footer\\"><button class=\\"btn btn-secondary\\" type=\\"button\\" data-dismiss=\\"modal\\">Cancel</button><button class=\\"btn btn-primary\\" type=\\"submit\\">Add</button></div></form></div></div></div></div></div></body><script src=\\"https://code.jquery.com/jquery-3.5.1.slim.min.js\\"></script><script src=\\"https://cdn.jsdelivr.net/npm/popper.js@1.16.1/dist/umd/popper.min.js\\"></script><script src=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js\\"></script><script src=\\"/javascripts/script.js\\"></script></html>"`;

exports[`Route GET /budgets/:id should render budget details page 1`] = `"<!DOCTYPE html><html><head><title>Personal Budget</title><link rel=\\"stylesheet\\" href=\\"/stylesheets/style.css\\"><link rel=\\"stylesheet\\" href=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.0/css/bootstrap.min.css\\" integrity=\\"sha384-9aIt2nRpC12Uk9gS9baDl411NQApFmC26EwAOH8WgZl5MYYxFfc+NcPb1dKGj7Sk\\" crossorigin=\\"anonymous\\"><link rel=\\"stylesheet\\" href=\\"https://use.fontawesome.com/releases/v5.7.0/css/all.css\\" crossorigin=\\"anonymous\\"></head><body class=\\"container\\"><nav class=\\"navbar navbar-expand-lg navbar-dark bg-dark\\"><a class=\\"navbar-brand\\" href=\\"#\\">Personal Budget</a><button class=\\"navbar-toggler\\" type=\\"button\\" data-toggle=\\"collapse\\" data-target=\\"#navbarSupportedContent\\" aria-controls=\\"navbarSupportedContent\\" aria-expanded=\\"false\\" aria-label=\\"Toggle navigation\\"><span class=\\"navbar-toggler-icon\\"></span></button><div class=\\"collapse navbar-collapse\\" id=\\"navbarSupportedContent\\"><ul class=\\"navbar-nav mr-auto\\"><li class=\\"nav-item\\"><a class=\\"nav-link\\" href=\\"/budgets\\">Budgets</a></li><li class=\\"nav-item\\"><a class=\\"nav-link\\" href=\\"/expenses\\">Expenses</a></li></ul></div></nav><div><h1 class=\\"mb-4 mt-3\\">Essence</h1></div><div class=\\"row\\"><div class=\\"col-12\\">Dépenses concernant l'essence des voitures</div><div class=\\"col-10\\">Available: 79€</div><button class=\\"col-1 btn btn-primary btn-sm\\" data-target=\\"#editBudgetModal\\" data-toggle=\\"modal\\"><span class=\\"fas fa-pencil-alt\\"></span></button><button class=\\"col-1 btn btn-danger btn-sm\\" onclick=\\"deleteBudgetLine(&quot;4&quot;)\\"><span class=\\"fas fa-trash-alt\\"></span></button><table class=\\"table col-12 mt-2\\"><thead><tr><th>Expense</th><th>Amount</th><th>Date</th></tr></thead><tbody></tbody><tr><td>Cmax</td><td>41 €</td><td>2020-08-29</td></tr><tr><td>Ka+</td><td>30 €</td><td>2020-08-22</td></tr></table></div><div class=\\"modal fade\\" id=\\"editBudgetModal\\"><div class=\\"modal-dialog\\"><div class=\\"modal-content\\"><div class=\\"modal-header\\"><h4 class=\\"modal-title\\">Edit</h4></div><div class=\\"modal-body\\"><form method=\\"POST\\" action=\\"/budgets/4\\"><div class=\\"form-group\\"><label for=\\"budgetNameEdit\\">Name</label><input class=\\"form-control\\" id=\\"budgetNameEdit\\" name=\\"name\\" type=\\"text\\" value=\\"Essence\\" aria-describedby=\\"budgetNameEditHelp\\" required><small class=\\"form-text text-muted\\" id=\\"budgetNameEditHelp\\">Budget line name must be unique</small><label for=\\"budgetAmountEdit\\">Amount (€)</label><input class=\\"form-control\\" id=\\"budgetAmountEdit\\" name=\\"amount\\" type=\\"number\\" step=\\"0.01\\" value=\\"150\\" aria-describedby=\\"budgetAmountEditHelp\\" required><small class=\\"form-text text-muted\\" id=\\"budgetAmountEditHelp\\">Budget line amount must be gretter than 0</small><label for=\\"budgetDescEdit\\">Description</label><textarea class=\\"form-control\\" id=\\"budgetDescEdit\\" name=\\"description\\">Dépenses concernant l'essence des voitures</textarea></div><div class=\\"modal-footer\\"><button class=\\"btn btn-secondary\\" type=\\"button\\" data-dismiss=\\"modal\\">Cancel</button><button class=\\"btn btn-primary\\" type=\\"submit\\">Edit</button></div></form></div></div></div></div></body><script src=\\"https://code.jquery.com/jquery-3.5.1.slim.min.js\\"></script><script src=\\"https://cdn.jsdelivr.net/npm/popper.js@1.16.1/dist/umd/popper.min.js\\"></script><script src=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js\\"></script><script src=\\"/javascripts/script.js\\"></script></html>"`;

exports[`Route GET /expenses should render expenses page 1`] = `"<!DOCTYPE html><html><head><title>Personal Budget</title><link rel=\\"stylesheet\\" href=\\"/stylesheets/style.css\\"><link rel=\\"stylesheet\\" href=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.0/css/bootstrap.min.css\\" integrity=\\"sha384-9aIt2nRpC12Uk9gS9baDl411NQApFmC26EwAOH8WgZl5MYYxFfc+NcPb1dKGj7Sk\\" crossorigin=\\"anonymous\\"><link rel=\\"stylesheet\\" href=\\"https://use.fontawesome.com/releases/v5.7.0/css/all.css\\" crossorigin=\\"anonymous\\"></head><body class=\\"container\\"><nav class=\\"navbar navbar-expand-lg navbar-dark bg-dark\\"><a class=\\"navbar-brand\\" href=\\"#\\">Personal Budget</a><button class=\\"navbar-toggler\\" type=\\"button\\" data-toggle=\\"collapse\\" data-target=\\"#navbarSupportedContent\\" aria-controls=\\"navbarSupportedContent\\" aria-expanded=\\"false\\" aria-label=\\"Toggle navigation\\"><span class=\\"navbar-toggler-icon\\"></span></button><div class=\\"collapse navbar-collapse\\" id=\\"navbarSupportedContent\\"><ul class=\\"navbar-nav mr-auto\\"><li class=\\"nav-item\\"><a class=\\"nav-link\\" href=\\"/budgets\\">Budgets</a></li><li class=\\"nav-item active\\"><a class=\\"nav-link\\" href=\\"/expenses\\">Expenses</a></li></ul></div></nav><div class=\\"row\\"><h1 class=\\"mb-4 mt-3 col-md-9\\">Expenses<button class=\\"btn btn-primary addBtn\\" type=\\"button\\" data-target=\\"#addExpenseModal\\" data-toggle=\\"modal\\"><span class=\\"fas fa-plus\\"></span></button></h1><div class=\\"col-md-3 mb-4 mt-3\\"><div class=\\"input-group\\" style=\\"padding-top: 4%\\"><input class=\\"form-control\\" id=\\"expense-filter\\" type=\\"text\\" placeholder=\\"Budget Name\\" aria-label=\\"Budget Name\\" aria-describedby=\\"budget-name-filter\\"><div class=\\"input-group-append\\"><button class=\\"btn btn-outline-secondary\\" id=\\"budget-name-filter\\" type=\\"button\\" onclick=\\"filterExpenseOnBudgetName()\\">Filter</button></div></div></div></div><table class=\\"table\\"><thead><tr><th>Expense</th><th>Amount</th><th>Budget Line</th><th>Date</th><th>Action</th></tr></thead><tbody></tbody><tr><td>Cmax</td><td>41 €</td><td>Essence</td><td>2020-08-29</td><td> <button class=\\"btn btn-danger btn-sm\\" onclick=\\"deleteExpense(&quot;100&quot;)\\"><span class=\\"fas fa-trash-alt\\"></span></button></td></tr><tr><td>Ka+</td><td>30 €</td><td>Essence</td><td>2020-08-22</td><td> <button class=\\"btn btn-danger btn-sm\\" onclick=\\"deleteExpense(&quot;101&quot;)\\"><span class=\\"fas fa-trash-alt\\"></span></button></td></tr><tr><td>courses</td><td>61.45 €</td><td>Course</td><td>2020-10-08</td><td> <button class=\\"btn btn-danger btn-sm\\" onclick=\\"deleteExpense(&quot;200&quot;)\\"><span class=\\"fas fa-trash-alt\\"></span></button></td></tr></table><div class=\\"modal fade\\" id=\\"addExpenseModal\\"><div class=\\"modal-dialog\\"><div class=\\"modal-content\\"><div class=\\"modal-header\\"><h4 class=\\"modal-title\\">Add Expense</h4></div><div class=\\"modal-body\\"><form method=\\"POST\\" action=\\"/expenses\\"><div class=\\"form-group\\"><label for=\\"expenseName\\">Name</label><input class=\\"form-control\\" id=\\"expenseName\\" name=\\"name\\" type=\\"text\\" required><label for=\\"expenseAmount\\">Amount (€)</label><input class=\\"form-control\\" id=\\"expenseAmount\\" name=\\"amount\\" type=\\"number\\" step=\\"0.01\\" aria-describedby=\\"expenseAmountHelp\\" required><small class=\\"form-text text-muted\\" id=\\"expenseAmountHelp\\">Expense amount must be gretter than 0</small><label for=\\"expenseDate\\">Date</label><input class=\\"form-control\\" id=\\"expenseDate\\" name=\\"date\\" type=\\"date\\" required><label for=\\"budgetLine\\">Budget Line</label><select class=\\"custom-select\\" id=\\"budgetLine\\" name=\\"budgetlineId\\"><option value=\\"1\\">Electricité</option><option value=\\"2\\">Crédit voiture</option><option value=\\"3\\">Crédit immobilier</option><option value=\\"4\\">Essence</option><option value=\\"5\\">Assurance voiture</option><option value=\\"6\\">Alimentation</option><option value=\\"7\\">Animaux</option><option value=\\"8\\">Enfants</option></select></div><div class=\\"modal-footer\\"><button class=\\"btn btn-secondary\\" type=\\"button\\" data-dismiss=\\"modal\\">Cancel</button><button class=\\"btn btn-primary\\" type=\\"submit\\">Add</button></div></form></div></div></div></div></body><script src=\\"https://code.jquery.com/jquery-3.5.1.slim.min.js\\"></script><script src=\\"https://cdn.jsdelivr.net/npm/popper.js@1.16.1/dist/umd/popper.min.js\\"></script><script src=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js\\"></script><script src=\\"/javascripts/script.js\\"></script></html>"`;

exports[`Route GET /expenses/filter should render expenses filtered page 1`] = `"<!DOCTYPE html><html><head><title>Personal Budget</title><link rel=\\"stylesheet\\" href=\\"/stylesheets/style.css\\"><link rel=\\"stylesheet\\" href=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.0/css/bootstrap.min.css\\" integrity=\\"sha384-9aIt2nRpC12Uk9gS9baDl411NQApFmC26EwAOH8WgZl5MYYxFfc+NcPb1dKGj7Sk\\" crossorigin=\\"anonymous\\"><link rel=\\"stylesheet\\" href=\\"https://use.fontawesome.com/releases/v5.7.0/css/all.css\\" crossorigin=\\"anonymous\\"></head><body class=\\"container\\"><nav class=\\"navbar navbar-expand-lg navbar-dark bg-dark\\"><a class=\\"navbar-brand\\" href=\\"#\\">Personal Budget</a><button class=\\"navbar-toggler\\" type=\\"button\\" data-toggle=\\"collapse\\" data-target=\\"#navbarSupportedContent\\" aria-controls=\\"navbarSupportedContent\\" aria-expanded=\\"false\\" aria-label=\\"Toggle navigation\\"><span class=\\"navbar-toggler-icon\\"></span></button><div class=\\"collapse navbar-collapse\\" id=\\"navbarSupportedContent\\"><ul class=\\"navbar-nav mr-auto\\"><li class=\\"nav-item\\"><a class=\\"nav-link\\" href=\\"/budgets\\">Budgets</a></li><li class=\\"nav-item active\\"><a class=\\"nav-link\\" href=\\"/expenses\\">Expenses</a></li></ul></div></nav><div class=\\"row\\"><h1 class=\\"mb-4 mt-3 col-md-9\\">Expenses<button class=\\"btn btn-primary addBtn\\" type=\\"button\\" data-target=\\"#addExpenseModal\\" data-toggle=\\"modal\\"><span class=\\"fas fa-plus\\"></span></button></h1><div class=\\"col-md-3 mb-4 mt-3\\"><div class=\\"input-group\\" style=\\"padding-top: 4%\\"><input class=\\"form-control\\" id=\\"expense-filter\\" type=\\"text\\" placeholder=\\"Budget Name\\" aria-label=\\"Budget Name\\" aria-describedby=\\"budget-name-filter\\"><div class=\\"input-group-append\\"><button class=\\"btn btn-outline-secondary\\" id=\\"budget-name-filter\\" type=\\"button\\" onclick=\\"filterExpenseOnBudgetName()\\">Filter</button></div></div></div></div><table class=\\"table\\"><thead><tr><th>Expense</th><th>Amount</th><th>Budget Line</th><th>Date</th><th>Action</th></tr></thead><tbody></tbody><tr><td>Cmax</td><td>41 €</td><td>Essence</td><td>2020-08-29</td><td> <button class=\\"btn btn-danger btn-sm\\" onclick=\\"deleteExpense(&quot;100&quot;)\\"><span class=\\"fas fa-trash-alt\\"></span></button></td></tr><tr><td>Ka+</td><td>30 €</td><td>Essence</td><td>2020-08-22</td><td> <button class=\\"btn btn-danger btn-sm\\" onclick=\\"deleteExpense(&quot;101&quot;)\\"><span class=\\"fas fa-trash-alt\\"></span></button></td></tr></table><div class=\\"modal fade\\" id=\\"addExpenseModal\\"><div class=\\"modal-dialog\\"><div class=\\"modal-content\\"><div class=\\"modal-header\\"><h4 class=\\"modal-title\\">Add Expense</h4></div><div class=\\"modal-body\\"><form method=\\"POST\\" action=\\"/expenses\\"><div class=\\"form-group\\"><label for=\\"expenseName\\">Name</label><input class=\\"form-control\\" id=\\"expenseName\\" name=\\"name\\" type=\\"text\\" required><label for=\\"expenseAmount\\">Amount (€)</label><input class=\\"form-control\\" id=\\"expenseAmount\\" name=\\"amount\\" type=\\"number\\" step=\\"0.01\\" aria-describedby=\\"expenseAmountHelp\\" required><small class=\\"form-text text-muted\\" id=\\"expenseAmountHelp\\">Expense amount must be gretter than 0</small><label for=\\"expenseDate\\">Date</label><input class=\\"form-control\\" id=\\"expenseDate\\" name=\\"date\\" type=\\"date\\" required><label for=\\"budgetLine\\">Budget Line</label><select class=\\"custom-select\\" id=\\"budgetLine\\" name=\\"budgetlineId\\"><option value=\\"1\\">Electricité</option><option value=\\"2\\">Crédit voiture</option><option value=\\"3\\">Crédit immobilier</option><option value=\\"4\\">Essence</option><option value=\\"5\\">Assurance voiture</option><option value=\\"6\\">Alimentation</option><option value=\\"7\\">Animaux</option><option value=\\"8\\">Enfants</option></select></div><div class=\\"modal-footer\\"><button class=\\"btn btn-secondary\\" type=\\"button\\" data-dismiss=\\"modal\\">Cancel</button><button class=\\"btn btn-primary\\" type=\\"submit\\">Add</button></div></form></div></div></div></div></body><script src=\\"https://code.jquery.com/jquery-3.5.1.slim.min.js\\"></script><script src=\\"https://cdn.jsdelivr.net/npm/popper.js@1.16.1/dist/umd/popper.min.js\\"></script><script src=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js\\"></script><script src=\\"/javascripts/script.js\\"></script></html>"`;

exports[`Route PATCH /budgets should render budget details page 1`] = `"<!DOCTYPE html><html><head><title>Personal Budget</title><link rel=\\"stylesheet\\" href=\\"/stylesheets/style.css\\"><link rel=\\"stylesheet\\" href=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.0/css/bootstrap.min.css\\" integrity=\\"sha384-9aIt2nRpC12Uk9gS9baDl411NQApFmC26EwAOH8WgZl5MYYxFfc+NcPb1dKGj7Sk\\" crossorigin=\\"anonymous\\"><link rel=\\"stylesheet\\" href=\\"https://use.fontawesome.com/releases/v5.7.0/css/all.css\\" crossorigin=\\"anonymous\\"></head><body class=\\"container\\"><nav class=\\"navbar navbar-expand-lg navbar-dark bg-dark\\"><a class=\\"navbar-brand\\" href=\\"#\\">Personal Budget</a><button class=\\"navbar-toggler\\" type=\\"button\\" data-toggle=\\"collapse\\" data-target=\\"#navbarSupportedContent\\" aria-controls=\\"navbarSupportedContent\\" aria-expanded=\\"false\\" aria-label=\\"Toggle navigation\\"><span class=\\"navbar-toggler-icon\\"></span></button><div class=\\"collapse navbar-collapse\\" id=\\"navbarSupportedContent\\"><ul class=\\"navbar-nav mr-auto\\"><li class=\\"nav-item\\"><a class=\\"nav-link\\" href=\\"/budgets\\">Budgets</a></li><li class=\\"nav-item\\"><a class=\\"nav-link\\" href=\\"/expenses\\">Expenses</a></li></ul></div></nav><div><h1 class=\\"mb-4 mt-3\\">Essence</h1></div><div class=\\"row\\"><div class=\\"col-12\\">Dépenses concernant l'essence des voitures</div><div class=\\"col-10\\">Available: 29€</div><button class=\\"col-1 btn btn-primary btn-sm\\" data-target=\\"#editBudgetModal\\" data-toggle=\\"modal\\"><span class=\\"fas fa-pencil-alt\\"></span></button><button class=\\"col-1 btn btn-danger btn-sm\\" onclick=\\"deleteBudgetLine(&quot;4&quot;)\\"><span class=\\"fas fa-trash-alt\\"></span></button><table class=\\"table col-12 mt-2\\"><thead><tr><th>Expense</th><th>Amount</th><th>Date</th></tr></thead><tbody></tbody><tr><td>Cmax</td><td>41 €</td><td>2020-08-29</td></tr><tr><td>Ka+</td><td>30 €</td><td>2020-08-22</td></tr></table></div><div class=\\"modal fade\\" id=\\"editBudgetModal\\"><div class=\\"modal-dialog\\"><div class=\\"modal-content\\"><div class=\\"modal-header\\"><h4 class=\\"modal-title\\">Edit</h4></div><div class=\\"modal-body\\"><form method=\\"POST\\" action=\\"/budgets/4\\"><div class=\\"form-group\\"><label for=\\"budgetNameEdit\\">Name</label><input class=\\"form-control\\" id=\\"budgetNameEdit\\" name=\\"name\\" type=\\"text\\" value=\\"Essence\\" aria-describedby=\\"budgetNameEditHelp\\" required><small class=\\"form-text text-muted\\" id=\\"budgetNameEditHelp\\">Budget line name must be unique</small><label for=\\"budgetAmountEdit\\">Amount (€)</label><input class=\\"form-control\\" id=\\"budgetAmountEdit\\" name=\\"amount\\" type=\\"number\\" step=\\"0.01\\" value=\\"100\\" aria-describedby=\\"budgetAmountEditHelp\\" required><small class=\\"form-text text-muted\\" id=\\"budgetAmountEditHelp\\">Budget line amount must be gretter than 0</small><label for=\\"budgetDescEdit\\">Description</label><textarea class=\\"form-control\\" id=\\"budgetDescEdit\\" name=\\"description\\">Dépenses concernant l'essence des voitures</textarea></div><div class=\\"modal-footer\\"><button class=\\"btn btn-secondary\\" type=\\"button\\" data-dismiss=\\"modal\\">Cancel</button><button class=\\"btn btn-primary\\" type=\\"submit\\">Edit</button></div></form></div></div></div></div></body><script src=\\"https://code.jquery.com/jquery-3.5.1.slim.min.js\\"></script><script src=\\"https://cdn.jsdelivr.net/npm/popper.js@1.16.1/dist/umd/popper.min.js\\"></script><script src=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js\\"></script><script src=\\"/javascripts/script.js\\"></script></html>"`;
